### PR TITLE
Refactor mr::DescriptorSet

### PR DESCRIPTION
**Pull request by cone-forest**
Prepare the module for experiments with DescriptorSet allocation strategies.
Allocate multiple sets from a pool.
Add ability to reset a pool.

### CODE REVIEW

File: src/renderer/resources/shaders/shader.hpp [hunk 1]
```diff
@@ -68,6 +83,42 @@ namespace mr {
...
+
+    return stage_bits[stage];
+  }
+
+  inline constexpr vk::ShaderStageFlagBits get_stage_flags(Shader::Stage stage) noexcept
+  {
+    return get_stage_flags(std::to_underlying(stage));
+  }
+
+  inline constexpr const char * get_stage_name(std::integral auto stage)
```
Comment (mitya-y) [inline src/renderer/resources/shaders/shader.hpp:105]: Make it noexcept as get_stage_flags method

File: src/renderer/resources/shaders/shader.cpp [hunk 2]
```diff
@@ -5,67 +5,50 @@
...
 
-/*
-mr::Shader::~Shader()
-{
-  /// TODO: destroy shader modules
-}
-*/
-
-void mr::Shader::compile(ShaderStages stage)
+void mr::Shader::compile(Shader::Stage stage)
```
Comment (mitya-y) [inline src/renderer/resources/shaders/shader.cpp:37]: Why this method and also load is not noexcept?

File: src/renderer/window/window_context.cpp [hunk 3]
```diff
@@ -255,7 +256,7 @@ void mr::WindowContext::create_framebuffers(const VulkanState &state)
 void mr::WindowContext::render()
 {
   static Shader _shader = Shader(_state, "default");
-  const std::vector<vk::VertexInputAttributeDescription> descrs {
+  static std::vector<vk::VertexInputAttributeDescription> descrs {
```
Comment (mitya-y) [inline src/renderer/window/window_context.cpp:259]: Fix code style, it looks awful

File: src/renderer/window/window_context.cpp [hunk 4]
```diff
@@ -364,18 +367,19 @@ void mr::WindowContext::render()
   for (unsigned i = 0; i < gbuffers_number; i++) {
     light_bindings[i].binding = i;
   }
+  std::vector<Shader::ResourceView> light_attach;
+  for (unsigned i = 0; i < gbuffers_number; i++) {
+    light_attach.emplace_back(0, 1, (Image *)&_gbuffers[i]);
+  }
+  static DescriptorSet light_set =
+    descriptor_alloc.allocate_set(Shader::Stage::Vertex, light_attach).value_or(DescriptorSet());
```
Comment (mitya-y) [inline src/renderer/window/window_context.cpp:375]: Light use GBuffers at fragment stage

File: src/renderer/resources/descriptor/descriptor.hpp [hunk 5]
```diff
@@ -2,74 +2,60 @@
...
-               uint set_number = 0);
+    std::optional<vk::UniqueDescriptorPool> allocate_pool(
+        std::span<vk::DescriptorPoolSize> sizes) noexcept;
 
-    void update_all_attachments(
-      const VulkanState &state,
-      const std::vector<Attachment::Data> &attachments);
+  public:
+    DescriptorAllocator(const VulkanState &state) : _state(state) {
+      // TODO: create pool
```
Comment (mitya-y) [inline src/renderer/resources/descriptor/descriptor.hpp:48]: Create pool, now render doesn't work

File: src/renderer/window/window_context.cpp [hunk 6]
```diff
@@ -364,18 +367,19 @@ void mr::WindowContext::render()
   for (unsigned i = 0; i < gbuffers_number; i++) {
     light_bindings[i].binding = i;
```
Comment (mitya-y) [inline src/renderer/window/window_context.cpp:368]: Now it variable isn't use, either remove it or maybe you forgot use it.

**Review by mitya-y — COMMENTED** [review]
Now in render assertion error.

File: src/renderer/resources/descriptor/descriptor.cpp [hunk 7]
```diff
@@ -1,19 +1,40 @@
...
-mr::Descriptor::Descriptor(const VulkanState &state, Pipeline *pipeline,
-                           const std::vector<Attachment::Data> &attachments,
-                           uint set_number)
-    : _set_number(set_number)
-    , _set_layout(pipeline->set_layout(_set_number))
+  return types[attachment.index()];
+}
+
+static std::vector<vk::DescriptorSetLayoutBinding>
+get_bindings(mr::Shader::Stage stage, std::span<mr::Shader::ResourceView> attachment_set)
```
[the remark this answers is not in the document]
Comment (dantibel) [reply src/renderer/resources/descriptor/descriptor.cpp:19]: It should be std::span<const T>, but some write methods cannot work with it (at least without ugly stuff like const_cast).

**Pull request merged**